# Pandas passo a passo — CSV — identificar valores ausentes

Arquivo: **roupas_nan.csv**. CSV é texto com campos separados por vírgulas neste exemplo. O arquivo usa UTF-8 e números inteiros. Cada linha é uma peça de roupa.

## Abrir o arquivo

No Colab: na barra esquerda, clique na pasta Arquivos e no botão de upload. Envie o arquivo de dados indicado abaixo. No VS Code: coloque o arquivo na mesma pasta do notebook e abra essa pasta como projeto.

Se surgir `FileNotFoundError`, confira nome e pasta. No Colab, os arquivos enviados podem desaparecer ao encerrar a sessão; envie novamente quando necessário.

Não execute todos os notebooks juntos: cada atividade usa sua própria variável `df`.

**Conceito novo:** valor ausente é uma informação que não foi registrada. `NaN` significa Not a Number e é uma representação comum de ausência em colunas numéricas. Dependendo do tipo e versão, o pandas também pode usar `pd.NA` ou `NaT`. Zero é um valor conhecido: não significa ausência. Use `isna()` para detectar; não compare com `== NaN`.

## Como usar


1. Na primeira parte, tente resolver sem olhar o gabarito.
2. Escreva o código nas células de código e suas explicações nas células de texto.
3. Execute as células de cima para baixo.

**Como pensar:** qual informação preciso? Qual coluna contém essa informação? Quero todas as linhas ou apenas algumas? Qual operação responde à pergunta?

Comentários iniciados por `#` não são executados. `NameError` pode indicar uma preparação não executada; `KeyError` pode indicar um nome de coluna escrito incorretamente.

## Preparação

`import pandas as pd` disponibiliza a biblioteca com o apelido `pd`. A função de leitura abre o arquivo e cria um DataFrame. `df =` guarda a tabela na variável `df`. O arquivo precisa estar na pasta de trabalho; o notebook não cria o arquivo de entrada.

Os números à esquerda são os índices das linhas. Eles começam em zero.

In [6]:
import pandas as pd

df = pd.read_csv("roupas_nan.csv", encoding="utf-8")
df

,peca,preco,estoque
0,Camiseta,40.0,20.0
1,Calça,120.0,8.0
2,Vestido,NaN,5.0
3,Jaqueta,200.0,NaN
4,Shorts,60.0,0.0
5,Saia,90.0,7.0


## Parte 1 — perguntas

Para cada questão: **preveja → escreva → execute → interprete**.

### Questão 1

Mostre as três primeiras linhas.

**Pense antes:** Qual peça tem preço vazio?

**Minha previsão:**

Escreva aqui.

In [7]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 2

Veja o tamanho, os nomes das colunas e um resumo da tabela.

**Pense antes:** Quantas linhas e colunas há?

**Minha previsão:**

Escreva aqui.

In [8]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 3

Mostre somente os preços.

**Pense antes:** O valor ausente é zero?

**Minha previsão:**

Escreva aqui.

In [9]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 4

Calcule a média dos preços registrados.

**Pense antes:** Dividiremos por cinco ou seis?

**Minha previsão:**

Escreva aqui.

In [10]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 5

Identifique os valores ausentes com isna.

**Pense antes:** O que True indica neste resultado?

**Minha previsão:**

Escreva aqui.

In [11]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 6

Conte os valores ausentes de cada coluna.

**Pense antes:** Por que somar True e False funciona?

**Minha previsão:**

Escreva aqui.

In [12]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 7

Mostre todas as informações das peças sem preço.

**Pense antes:** Como usar a condição de ausência para filtrar?

**Minha previsão:**

Escreva aqui.

In [13]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 8

Mostre os nomes das peças com preço registrado usando loc.

**Pense antes:** Qual operação é o contrário de isna?

**Minha previsão:**

Escreva aqui.

In [14]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 9

Compare estoque zero com estoque ausente.

**Pense antes:** Shorts e Jaqueta têm o mesmo problema?

**Minha previsão:**

Escreva aqui.

In [15]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

## Parte 2 — resolução comentada

Compare com suas tentativas. Leia como pensar, execute o código e explique o resultado com suas palavras. Comece executando a preparação novamente.

In [16]:
import pandas as pd

df = pd.read_csv("roupas_nan.csv", encoding="utf-8")
df

,peca,preco,estoque
0,Camiseta,40.0,20.0
1,Calça,120.0,8.0
2,Vestido,NaN,5.0
3,Jaqueta,200.0,NaN
4,Shorts,60.0,0.0
5,Saia,90.0,7.0


### Resolução 1

**Pedido:** Mostre as três primeiras linhas.

**Como pensar:** Primeiro visualizamos o arquivo que acabamos de abrir.

In [17]:
df.head(3)

,peca,preco,estoque
0,Camiseta,40.0,20.0
1,Calça,120.0,8.0
2,Vestido,NaN,5.0


**Resultado explicado:** Camiseta, Calça e Vestido aparecem; o preço do Vestido está ausente.

### Resolução 2

**Pedido:** Veja o tamanho, os nomes das colunas e um resumo da tabela.

**Como pensar:** shape mostra linhas e colunas. columns lista os nomes. info informa tipos e quantidades preenchidas.

In [18]:
print(df.shape)
print(df.columns)
df.info()

(6, 3)
Index(['peca', 'preco', 'estoque'], dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   peca     6 non-null      object 
 1   preco    5 non-null      float64
 2   estoque  5 non-null      float64
dtypes: float64(2), object(1)
memory usage: 272.0+ bytes


**Resultado explicado:** São 6 linhas e 3 colunas; preco e estoque têm 5 valores preenchidos cada.

### Resolução 3

**Pedido:** Mostre somente os preços.

**Como pensar:** Selecionamos a coluna antes de fazer cálculos.

In [19]:
df["preco"]

0     40.0
1    120.0
2      NaN
3    200.0
4     60.0
5     90.0
Name: preco, dtype: float64

**Resultado explicado:** O Vestido não tem preço registrado. Não podemos concluir que ele é gratuito.

### Resolução 4

**Pedido:** Calcule a média dos preços registrados.

**Como pensar:** mean ignora os valores ausentes por padrão.

In [20]:
df["preco"].mean()

np.float64(102.0)

**Resultado explicado:** A soma dos cinco preços é 510; a média é 102. Não é a média de todos os produtos, pois falta um preço.

### Resolução 5

**Pedido:** Identifique os valores ausentes com isna.

**Como pensar:** isna verifica cada célula; True indica ausência.

In [21]:
df.isna()

,peca,preco,estoque
0,False,False,False
1,False,False,False
2,False,True,False
3,False,False,True
4,False,False,False
5,False,False,False


**Resultado explicado:** True aparece no preço do Vestido e no estoque da Jaqueta.

### Resolução 6

**Pedido:** Conte os valores ausentes de cada coluna.

**Como pensar:** Na soma, True conta como 1 e False como 0; a soma padrão é por coluna.

In [22]:
df.isna().sum()

peca       0
preco      1
estoque    1
dtype: int64

**Resultado explicado:** peca: 0; preco: 1; estoque: 1.

### Resolução 7

**Pedido:** Mostre todas as informações das peças sem preço.

**Como pensar:** O resultado booleano de isna seleciona as linhas ausentes.

In [23]:
df[df["preco"].isna()]

,peca,preco,estoque
2,Vestido,NaN,5.0


**Resultado explicado:** Somente Vestido aparece.

### Resolução 8

**Pedido:** Mostre os nomes das peças com preço registrado usando loc.

**Como pensar:** notna identifica preenchidos; loc seleciona essas linhas e a coluna peca.

In [24]:
df.loc[df["preco"].notna(), "peca"]

0    Camiseta
1       Calça
3     Jaqueta
4      Shorts
5        Saia
Name: peca, dtype: object

**Resultado explicado:** Camiseta, Calça, Jaqueta, Shorts e Saia.

### Resolução 9

**Pedido:** Compare estoque zero com estoque ausente.

**Como pensar:** Zero significa estoque registrado sem unidades; ausência significa quantidade desconhecida.

In [25]:
print("Estoque zero:")
print(df.loc[df["estoque"] == 0, "peca"])
print("Estoque ausente:")
print(df.loc[df["estoque"].isna(), "peca"])

Estoque zero:
4    Shorts
Name: peca, dtype: object
Estoque ausente:
3    Jaqueta
Name: peca, dtype: object


**Resultado explicado:** Shorts tem zero unidades. O estoque da Jaqueta é desconhecido.

## Lembretes

| Recurso | Significado |
|---|---|
| `df["coluna"]` | Seleciona uma coluna |
| `df[condicao]` | Seleciona as linhas que atendem à condição |
| `df.loc[linhas, colunas]` | Seleciona por rótulos ou condições |
| `df.iloc[linhas, colunas]` | Seleciona por posições começando em zero |
| `:` sozinho | Todas as linhas ou colunas |
| `0:3` no iloc | Posições 0, 1 e 2; exclui a posição 3 |
| `==` | Compara valores |
| `=` | Atribui um valor |

Selecionar ou mostrar um resultado não significa apagar as outras informações. A alteração de valores é feita explicitamente, como na questão 9.

Documentação: https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html